Userland/Libraries/LibGfx/BitmapSequence.cpp

```diff
@@ -0,0 +1,129 @@
+    : m_bitmaps(move(bitmaps))
+{
+}
+
+}
+
+namespace IPC {
+
+template<>
+ErrorOr<void> encode(Encoder& encoder, Gfx::BitmapSequence const& bitmap_sequence)
+{
+    Vector<Optional<NonnullRefPtr<Gfx::Bitmap>>> bitmaps = bitmap_sequence.bitmaps();
+
+    TRY(encoder.encode_size(bitmaps.size()));
+
+    size_t total_buffer_size = 0;
+
+    for (auto const& bitmap_option : bitmaps) {
+        TRY(encoder.encode(bitmap_option.has_value()));
+
+        if (bitmap_option.has_value()) {
+            auto bitmap = bitmap_option.value();
+            TRY(encoder.encode(static_cast<u32>(bitmap->format())));
+            TRY(encoder.encode(static_cast<u32>(bitmap->alpha_type())));
+            TRY(encoder.encode(bitmap->size_in_bytes()));
+            TRY(encoder.encode(bitmap->size()));
+
+            total_buffer_size += bitmap_option.value()->size_in_bytes();
+        }
+    }
+
+    // collate all of the bitmap data into one contiguous buffer
+    auto collated_buffer = TRY(Core::AnonymousBuffer::create_with_size(total_buffer_size));
+
+    auto* write_pointer = collated_buffer.data<u8>();
+    for (auto const& bitmap_option : bitmaps) {
+        if (bitmap_option.has_value()) {
+            auto bitmap = bitmap_option.value();
+            memcpy(write_pointer, bitmap->scanline(0), bitmap->size_in_bytes());
+            write_pointer += bitmap->size_in_bytes();
+        }
+    }
+
+    TRY(encoder.encode(collated_buffer));
+
+    return {};
+}
+
+template<>
+ErrorOr<Gfx::BitmapSequence> decode(Decoder& decoder)
+{
+    // a struct to temporarily store bitmap fields before the buffer data is decoded
+    struct BitmapMetadata {
+        Gfx::BitmapFormat format;
+        Gfx::AlphaType alpha_type;
+        Gfx::IntSize size;
+        size_t size_in_bytes;
+    };
+
+    Vector<Optional<BitmapMetadata>> bitmaps_metadata;
+
+    size_t num_bitmaps = TRY(decoder.decode_size());
+
+    for (size_t i = 0; i < num_bitmaps; i++) {
+        auto valid = TRY(decoder.decode<bool>());
+
+        if (valid) {
+            auto raw_bitmap_format = TRY(decoder.decode<u32>());
+            if (!Gfx::is_valid_bitmap_format(raw_bitmap_format))
+                return Error::from_string_literal("IPC: Invalid Gfx::BitmapSequence format");
+            auto format = static_cast<Gfx::BitmapFormat>(raw_bitmap_format);
+
+            auto raw_alpha_type = TRY(decoder.decode<u32>());
+            if (!Gfx::is_valid_alpha_type(raw_alpha_type))
+                return Error::from_string_literal("IPC: Invalid Gfx::BitmapSequence alpha type");
+            auto alpha_type = static_cast<Gfx::AlphaType>(raw_alpha_type);
+
+            auto size_in_bytes = TRY(decoder.decode<size_t>());
+            auto size = TRY(decoder.decode<Gfx::IntSize>());
+
+            bitmaps_metadata.append(BitmapMetadata { format, alpha_type, size, size_in_bytes });
+        } else {
+            bitmaps_metadata.append({});
+        }
+    }
```

Comment:
This could be factored out into a decode method for BitmapMetadata. Using the other suggestions in the file, then you could just `auto bitmaps_metadata = TRY(decoder.decode<Vector<Optional<BitmapMetadata>>>());`